{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import dill\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from freqs import Freqs, load_freqs, dump_freqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%aimport freqs\n",
    "%autoreload 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## モジュールの動作確認&pressureの算出（02/04）"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "length error: not multiple of 3 (1357)\n",
      "length error: not multiple of 3 (623)\n",
      "stop error: inframe stop codon found\n",
      "length error: not multiple of 3 (947)\n",
      "length error: not multiple of 3 (407)\n",
      "length error: not multiple of 3 (812)\n"
     ]
    }
   ],
   "source": [
    "cds_fp = \"./data/GCF_900128725.1_BCifornacula_v1.0_cds_from_genomic.fna\"\n",
    "cdss = list(SeqIO.parse(cds_fp, 'fasta'))\n",
    "freqs = Freqs()\n",
    "for cds in cdss:\n",
    "    freqs.update(cds.seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## freqsからそれぞれのcodonの確率を算出（02/04）"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "freqs_fp = \"./data/freqs/GCF_000010525.1.freqs\"\n",
    "freqs = load_freqs(freqs_fp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4723"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "freqs.aa['*']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import itertools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AAA': 0.07672873788522269,\n",
       " 'AAC': 0.6730963359070199,\n",
       " 'AAG': 0.9232712621147773,\n",
       " 'AAT': 0.3269036640929801,\n",
       " 'ACA': 0.033610214709274555,\n",
       " 'ACC': 0.5788878921677132,\n",
       " 'ACG': 0.36385241795496115,\n",
       " 'ACT': 0.023649475168051073,\n",
       " 'AGA': 0.006185030251889621,\n",
       " 'AGC': 0.2948471999902247,\n",
       " 'AGG': 0.02073645846424532,\n",
       " 'AGT': 0.02265423575556886,\n",
       " 'ATA': 0.012407319229984354,\n",
       " 'ATC': 0.893612679409564,\n",
       " 'ATG': 1.0,\n",
       " 'ATT': 0.09398000136045168,\n",
       " 'CAA': 0.07281771983618157,\n",
       " 'CAC': 0.5645443443245206,\n",
       " 'CAG': 0.9271822801638184,\n",
       " 'CAT': 0.43545565567547945,\n",
       " 'CCA': 0.029226178272853596,\n",
       " 'CCC': 0.401273377525684,\n",
       " 'CCG': 0.5217772614410351,\n",
       " 'CCT': 0.04772318276042734,\n",
       " 'CGA': 0.020942925366703172,\n",
       " 'CGC': 0.627031005942656,\n",
       " 'CGG': 0.2481821935761863,\n",
       " 'CGT': 0.07692238639831954,\n",
       " 'CTA': 0.006056307289393278,\n",
       " 'CTC': 0.40847640525728696,\n",
       " 'CTG': 0.47519278335515786,\n",
       " 'CTT': 0.0736092924002134,\n",
       " 'GAA': 0.2870973880774537,\n",
       " 'GAC': 0.6791865108000192,\n",
       " 'GAG': 0.7129026119225462,\n",
       " 'GAT': 0.32081348919998076,\n",
       " 'GCA': 0.04314431076681684,\n",
       " 'GCC': 0.5336900505949724,\n",
       " 'GCG': 0.37828785109951285,\n",
       " 'GCT': 0.04487778753869787,\n",
       " 'GGA': 0.04788481584433635,\n",
       " 'GGC': 0.7422399814686125,\n",
       " 'GGG': 0.13336084086170952,\n",
       " 'GGT': 0.07651436182534167,\n",
       " 'GTA': 0.01549959830883964,\n",
       " 'GTC': 0.32859148414740036,\n",
       " 'GTG': 0.6047845880433989,\n",
       " 'GTT': 0.05112432950036112,\n",
       " 'TAA': 0.059707812830827864,\n",
       " 'TAC': 0.38727070092472665,\n",
       " 'TAG': 0.0942197755663773,\n",
       " 'TAT': 0.6127292990752734,\n",
       " 'TCA': 0.021407886215618472,\n",
       " 'TCC': 0.3669277483840223,\n",
       " 'TCG': 0.2639695010936106,\n",
       " 'TCT': 0.030193428560955044,\n",
       " 'TGA': 0.8460724116027948,\n",
       " 'TGC': 0.9132268266969427,\n",
       " 'TGG': 1.0,\n",
       " 'TGT': 0.08677317330305734,\n",
       " 'TTA': 0.0019217711819195886,\n",
       " 'TTC': 0.9008232547525847,\n",
       " 'TTG': 0.03474344051602891,\n",
       " 'TTT': 0.09917674524741527}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codons = [''.join(_) for _ in itertools.product(\"TCAG\", repeat=3)]\n",
    "\n",
    "probs = dict()\n",
    "for codon in codons:\n",
    "            aa = str(Seq(codon).translate(table=11))\n",
    "            probs[codon] = freqs.codon[codon] / freqs.aa[aa] if freqs.aa[aa] > 0 else 0\n",
    "probs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "dct_lst = []\n",
    "for c1 in codons:\n",
    "    for c2 in codons:\n",
    "        bicodon = c1 + c2\n",
    "        biaa = str(Seq(bicodon).translate(table=11))\n",
    "        a1 = biaa[0]\n",
    "        a2 = biaa[1]\n",
    "        \n",
    "        dct = {\n",
    "            \"bicodon\": bicodon,\n",
    "            \"freq\": freqs.bicodon[c1][c2],\n",
    "            \"pred\": freqs.biaa[a1][a2] * probs[c1] * probs[c2]\n",
    "        }\n",
    "        dct_lst.append(dct)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bicodon</th>\n",
       "      <th>freq</th>\n",
       "      <th>pred</th>\n",
       "      <th>pres</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2603</th>\n",
       "      <td>AATAAG</td>\n",
       "      <td>18</td>\n",
       "      <td>281.296947</td>\n",
       "      <td>0.063989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1069</th>\n",
       "      <td>CTTAGC</td>\n",
       "      <td>15</td>\n",
       "      <td>232.205680</td>\n",
       "      <td>0.064598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2083</th>\n",
       "      <td>ATTATG</td>\n",
       "      <td>7</td>\n",
       "      <td>108.358942</td>\n",
       "      <td>0.064600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>TTAGTG</td>\n",
       "      <td>1</td>\n",
       "      <td>15.275552</td>\n",
       "      <td>0.065464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2093</th>\n",
       "      <td>ATTAGC</td>\n",
       "      <td>8</td>\n",
       "      <td>120.675919</td>\n",
       "      <td>0.066293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3529</th>\n",
       "      <td>GCGTAC</td>\n",
       "      <td>43</td>\n",
       "      <td>606.655677</td>\n",
       "      <td>0.070880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3528</th>\n",
       "      <td>GCGTAT</td>\n",
       "      <td>70</td>\n",
       "      <td>959.834314</td>\n",
       "      <td>0.072929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>456</th>\n",
       "      <td>TCGTAT</td>\n",
       "      <td>25</td>\n",
       "      <td>316.367053</td>\n",
       "      <td>0.079022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2091</th>\n",
       "      <td>ATTAAG</td>\n",
       "      <td>13</td>\n",
       "      <td>155.490110</td>\n",
       "      <td>0.083607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2504</th>\n",
       "      <td>ACGTAT</td>\n",
       "      <td>39</td>\n",
       "      <td>454.134966</td>\n",
       "      <td>0.085878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3627</th>\n",
       "      <td>GATAAG</td>\n",
       "      <td>64</td>\n",
       "      <td>673.850166</td>\n",
       "      <td>0.094977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3113</th>\n",
       "      <td>GTTAAC</td>\n",
       "      <td>10</td>\n",
       "      <td>104.852142</td>\n",
       "      <td>0.095372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1480</th>\n",
       "      <td>CCGTAT</td>\n",
       "      <td>60</td>\n",
       "      <td>619.594522</td>\n",
       "      <td>0.096838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2505</th>\n",
       "      <td>ACGTAC</td>\n",
       "      <td>29</td>\n",
       "      <td>287.032409</td>\n",
       "      <td>0.101034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3117</th>\n",
       "      <td>GTTAGC</td>\n",
       "      <td>11</td>\n",
       "      <td>108.486609</td>\n",
       "      <td>0.101395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3272</th>\n",
       "      <td>GTGTAT</td>\n",
       "      <td>82</td>\n",
       "      <td>806.358659</td>\n",
       "      <td>0.101692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>331</th>\n",
       "      <td>TCCTAG</td>\n",
       "      <td>1</td>\n",
       "      <td>9.783834</td>\n",
       "      <td>0.102209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1224</th>\n",
       "      <td>CTGTAT</td>\n",
       "      <td>107</td>\n",
       "      <td>940.752632</td>\n",
       "      <td>0.113739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2339</th>\n",
       "      <td>ACTATG</td>\n",
       "      <td>4</td>\n",
       "      <td>34.528234</td>\n",
       "      <td>0.115847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3116</th>\n",
       "      <td>GTTAGT</td>\n",
       "      <td>1</td>\n",
       "      <td>8.335440</td>\n",
       "      <td>0.119970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>457</th>\n",
       "      <td>TCGTAC</td>\n",
       "      <td>24</td>\n",
       "      <td>199.957291</td>\n",
       "      <td>0.120026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1065</th>\n",
       "      <td>CTTAAC</td>\n",
       "      <td>23</td>\n",
       "      <td>188.176259</td>\n",
       "      <td>0.122226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301</th>\n",
       "      <td>TCTAGC</td>\n",
       "      <td>5</td>\n",
       "      <td>40.648577</td>\n",
       "      <td>0.123006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3658</th>\n",
       "      <td>GACTAA</td>\n",
       "      <td>1</td>\n",
       "      <td>8.069995</td>\n",
       "      <td>0.123916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3782</th>\n",
       "      <td>GAGTCA</td>\n",
       "      <td>6</td>\n",
       "      <td>46.990891</td>\n",
       "      <td>0.127684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3107</th>\n",
       "      <td>GTTATG</td>\n",
       "      <td>18</td>\n",
       "      <td>139.467171</td>\n",
       "      <td>0.129063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>TTTAGC</td>\n",
       "      <td>13</td>\n",
       "      <td>100.387737</td>\n",
       "      <td>0.129498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>TTTAGT</td>\n",
       "      <td>1</td>\n",
       "      <td>7.713173</td>\n",
       "      <td>0.129648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>555</th>\n",
       "      <td>TATAAG</td>\n",
       "      <td>66</td>\n",
       "      <td>507.446672</td>\n",
       "      <td>0.130063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>TTGTAT</td>\n",
       "      <td>9</td>\n",
       "      <td>68.782575</td>\n",
       "      <td>0.130847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2956</th>\n",
       "      <td>AGATGT</td>\n",
       "      <td>4</td>\n",
       "      <td>0.419695</td>\n",
       "      <td>9.530725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1667</th>\n",
       "      <td>CAATTG</td>\n",
       "      <td>101</td>\n",
       "      <td>10.534662</td>\n",
       "      <td>9.587398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1422</th>\n",
       "      <td>CCATGA</td>\n",
       "      <td>73</td>\n",
       "      <td>7.566604</td>\n",
       "      <td>9.647657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3211</th>\n",
       "      <td>GTATAG</td>\n",
       "      <td>3</td>\n",
       "      <td>0.305217</td>\n",
       "      <td>9.829071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2194</th>\n",
       "      <td>ATACTA</td>\n",
       "      <td>5</td>\n",
       "      <td>0.501652</td>\n",
       "      <td>9.967077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2179</th>\n",
       "      <td>ATATTG</td>\n",
       "      <td>30</td>\n",
       "      <td>2.877843</td>\n",
       "      <td>10.424474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1162</th>\n",
       "      <td>CTATAA</td>\n",
       "      <td>1</td>\n",
       "      <td>0.094018</td>\n",
       "      <td>10.636227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2315</th>\n",
       "      <td>ACTTAG</td>\n",
       "      <td>3</td>\n",
       "      <td>0.271846</td>\n",
       "      <td>11.035648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2698</th>\n",
       "      <td>AAATAA</td>\n",
       "      <td>9</td>\n",
       "      <td>0.815472</td>\n",
       "      <td>11.036549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1674</th>\n",
       "      <td>CAATAA</td>\n",
       "      <td>8</td>\n",
       "      <td>0.695646</td>\n",
       "      <td>11.500104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2222</th>\n",
       "      <td>ATAAGA</td>\n",
       "      <td>4</td>\n",
       "      <td>0.339036</td>\n",
       "      <td>11.798166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2434</th>\n",
       "      <td>ACATTA</td>\n",
       "      <td>8</td>\n",
       "      <td>0.667226</td>\n",
       "      <td>11.989932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2946</th>\n",
       "      <td>AGATTA</td>\n",
       "      <td>2</td>\n",
       "      <td>0.156898</td>\n",
       "      <td>12.747134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1418</th>\n",
       "      <td>CCATAA</td>\n",
       "      <td>7</td>\n",
       "      <td>0.533980</td>\n",
       "      <td>13.109116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>TTACGA</td>\n",
       "      <td>6</td>\n",
       "      <td>0.428596</td>\n",
       "      <td>13.999206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2955</th>\n",
       "      <td>AGATAG</td>\n",
       "      <td>4</td>\n",
       "      <td>0.284966</td>\n",
       "      <td>14.036772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>TTAAGG</td>\n",
       "      <td>6</td>\n",
       "      <td>0.424370</td>\n",
       "      <td>14.138592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1930</th>\n",
       "      <td>CGATAA</td>\n",
       "      <td>9</td>\n",
       "      <td>0.611473</td>\n",
       "      <td>14.718554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1675</th>\n",
       "      <td>CAATAG</td>\n",
       "      <td>17</td>\n",
       "      <td>1.097739</td>\n",
       "      <td>15.486376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>TTAAGA</td>\n",
       "      <td>2</td>\n",
       "      <td>0.126576</td>\n",
       "      <td>15.800749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1154</th>\n",
       "      <td>CTATTA</td>\n",
       "      <td>3</td>\n",
       "      <td>0.187106</td>\n",
       "      <td>16.033697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2699</th>\n",
       "      <td>AAATAG</td>\n",
       "      <td>21</td>\n",
       "      <td>1.286827</td>\n",
       "      <td>16.319212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1931</th>\n",
       "      <td>CGATAG</td>\n",
       "      <td>16</td>\n",
       "      <td>0.964913</td>\n",
       "      <td>16.581802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2954</th>\n",
       "      <td>AGATAA</td>\n",
       "      <td>3</td>\n",
       "      <td>0.180585</td>\n",
       "      <td>16.612669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2990</th>\n",
       "      <td>AGAAGA</td>\n",
       "      <td>7</td>\n",
       "      <td>0.356303</td>\n",
       "      <td>19.646181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>TTAATA</td>\n",
       "      <td>4</td>\n",
       "      <td>0.172631</td>\n",
       "      <td>23.170841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3210</th>\n",
       "      <td>GTATAA</td>\n",
       "      <td>5</td>\n",
       "      <td>0.193418</td>\n",
       "      <td>25.850688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>TTATTA</td>\n",
       "      <td>2</td>\n",
       "      <td>0.059372</td>\n",
       "      <td>33.685938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1163</th>\n",
       "      <td>CTATAG</td>\n",
       "      <td>5</td>\n",
       "      <td>0.148362</td>\n",
       "      <td>33.701303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2178</th>\n",
       "      <td>ATATTA</td>\n",
       "      <td>7</td>\n",
       "      <td>0.159183</td>\n",
       "      <td>43.974618</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3877 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     bicodon  freq        pred       pres\n",
       "2603  AATAAG    18  281.296947   0.063989\n",
       "1069  CTTAGC    15  232.205680   0.064598\n",
       "2083  ATTATG     7  108.358942   0.064600\n",
       "179   TTAGTG     1   15.275552   0.065464\n",
       "2093  ATTAGC     8  120.675919   0.066293\n",
       "3529  GCGTAC    43  606.655677   0.070880\n",
       "3528  GCGTAT    70  959.834314   0.072929\n",
       "456   TCGTAT    25  316.367053   0.079022\n",
       "2091  ATTAAG    13  155.490110   0.083607\n",
       "2504  ACGTAT    39  454.134966   0.085878\n",
       "3627  GATAAG    64  673.850166   0.094977\n",
       "3113  GTTAAC    10  104.852142   0.095372\n",
       "1480  CCGTAT    60  619.594522   0.096838\n",
       "2505  ACGTAC    29  287.032409   0.101034\n",
       "3117  GTTAGC    11  108.486609   0.101395\n",
       "3272  GTGTAT    82  806.358659   0.101692\n",
       "331   TCCTAG     1    9.783834   0.102209\n",
       "1224  CTGTAT   107  940.752632   0.113739\n",
       "2339  ACTATG     4   34.528234   0.115847\n",
       "3116  GTTAGT     1    8.335440   0.119970\n",
       "457   TCGTAC    24  199.957291   0.120026\n",
       "1065  CTTAAC    23  188.176259   0.122226\n",
       "301   TCTAGC     5   40.648577   0.123006\n",
       "3658  GACTAA     1    8.069995   0.123916\n",
       "3782  GAGTCA     6   46.990891   0.127684\n",
       "3107  GTTATG    18  139.467171   0.129063\n",
       "45    TTTAGC    13  100.387737   0.129498\n",
       "44    TTTAGT     1    7.713173   0.129648\n",
       "555   TATAAG    66  507.446672   0.130063\n",
       "200   TTGTAT     9   68.782575   0.130847\n",
       "...      ...   ...         ...        ...\n",
       "2956  AGATGT     4    0.419695   9.530725\n",
       "1667  CAATTG   101   10.534662   9.587398\n",
       "1422  CCATGA    73    7.566604   9.647657\n",
       "3211  GTATAG     3    0.305217   9.829071\n",
       "2194  ATACTA     5    0.501652   9.967077\n",
       "2179  ATATTG    30    2.877843  10.424474\n",
       "1162  CTATAA     1    0.094018  10.636227\n",
       "2315  ACTTAG     3    0.271846  11.035648\n",
       "2698  AAATAA     9    0.815472  11.036549\n",
       "1674  CAATAA     8    0.695646  11.500104\n",
       "2222  ATAAGA     4    0.339036  11.798166\n",
       "2434  ACATTA     8    0.667226  11.989932\n",
       "2946  AGATTA     2    0.156898  12.747134\n",
       "1418  CCATAA     7    0.533980  13.109116\n",
       "158   TTACGA     6    0.428596  13.999206\n",
       "2955  AGATAG     4    0.284966  14.036772\n",
       "175   TTAAGG     6    0.424370  14.138592\n",
       "1930  CGATAA     9    0.611473  14.718554\n",
       "1675  CAATAG    17    1.097739  15.486376\n",
       "174   TTAAGA     2    0.126576  15.800749\n",
       "1154  CTATTA     3    0.187106  16.033697\n",
       "2699  AAATAG    21    1.286827  16.319212\n",
       "1931  CGATAG    16    0.964913  16.581802\n",
       "2954  AGATAA     3    0.180585  16.612669\n",
       "2990  AGAAGA     7    0.356303  19.646181\n",
       "162   TTAATA     4    0.172631  23.170841\n",
       "3210  GTATAA     5    0.193418  25.850688\n",
       "130   TTATTA     2    0.059372  33.685938\n",
       "1163  CTATAG     5    0.148362  33.701303\n",
       "2178  ATATTA     7    0.159183  43.974618\n",
       "\n",
       "[3877 rows x 4 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pres_df = pd.DataFrame(dct_lst)\n",
    "pres_df[\"pres\"] = pres_df[\"freq\"] / pres_df[\"pred\"]\n",
    "pres_df[pres_df[\"pres\"] > 0].sort_values(by=\"pres\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## cdsファイルから各frequenciesを算出する（02/03）"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "class Freqs:\n",
    "    def __init__(self):\n",
    "        self.codon = Counter()\n",
    "        self.aa = Counter()\n",
    "        self.bicodon = defaultdict(lambda: Counter())\n",
    "        self.biaa = defaultdict(lambda: Counter())\n",
    "        \n",
    "    def _is_typical(self, cds_seq):\n",
    "        \"\"\"\n",
    "        Check the following:\n",
    "        * length is multiple of 3\n",
    "        * Only ATGC used\n",
    "        * No premature stop codon included\n",
    "        \"\"\"\n",
    "        \n",
    "        if len(cds_seq) % 3 != 0:\n",
    "            print(\"length error\")\n",
    "            return False\n",
    "        \n",
    "        if not(set(cds_seq) <= set(\"ATGC\")):\n",
    "            print(\"character error: unknown characters ({}) included\".format(','.join(set(cds_seq) - set(\"ATGC\"))))\n",
    "            return False\n",
    "        \n",
    "        pro = cds_seq.translate(table=11)\n",
    "        if \"*\" in pro[:-1]:\n",
    "            print(\"stop error: inframe stop codon found\")\n",
    "            return False\n",
    "        \n",
    "        return True\n",
    "        \n",
    "    def update(self, cds_seq):\n",
    "        \"\"\"\n",
    "        Args:\n",
    "            cds_seq: Bio.Seq object\n",
    "        \"\"\"\n",
    "        \n",
    "        if not(self._is_typical(cds_seq)):\n",
    "            return False\n",
    "    \n",
    "        pro_seq = cds_seq.translate(table=11)\n",
    "        codons = [str(cds_seq[3*i:3*(i+1)]) for i in range(int(len(cds_seq)/3))]\n",
    "        aas = list(cds.seq.translate())\n",
    "        \n",
    "        self.codon.update(codons)\n",
    "        self.aa.update(aas)\n",
    "        for i in range(len(codons)-1):\n",
    "            self.bicodon[codons[i]][codons[i+1]] += 1\n",
    "        for i in range(len(aas) - 1):\n",
    "            self.biaa[aas[i]][aas[i+1]] += 1\n",
    "\n",
    "        return True\n",
    "    \n",
    "\n",
    "def dump_freqs(freqs, fp):\n",
    "    dill.dump(freqs, open(fp, \"wb\"))\n",
    "    \n",
    "def load_freqs(fp):\n",
    "    return dill.load(open(fp, \"rb\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "383"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds_fp = \"./data/GCF_900128725.1_BCifornacula_v1.0_cds_from_genomic.fna\"\n",
    "cdss = list(SeqIO.parse(cds_fp, 'fasta'))\n",
    "len(cdss)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "length error\n",
      "length error\n",
      "stop error: inframe stop codon found\n",
      "length error\n",
      "length error\n",
      "length error\n"
     ]
    }
   ],
   "source": [
    "freqs = Freqs()\n",
    "for cds in cdss:\n",
    "    freqs.update(cds.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fp = \"tmp.freqs\"\n",
    "dump_freqs(freqs, fp)\n",
    "freqs_new = load_freqs(fp)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
